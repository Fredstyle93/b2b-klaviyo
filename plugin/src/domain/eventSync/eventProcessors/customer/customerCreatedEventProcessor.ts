import { AbstractEventProcessor } from '../abstractEventProcessor';
import logger from '../../../../utils/log';
import { Customer, CustomerCreatedMessage } from '@commercetools/platform-sdk';
import config from "config";
export class CustomerCreatedEventProcessor extends AbstractEventProcessor {
    private readonly PROCESSOR_NAME = 'CustomerCreated';
    isEventValid(): boolean {
        const customerCreatedMessage = this.ctMessage as unknown as CustomerCreatedMessage;
        return (
            customerCreatedMessage.resource.typeId === 'customer' &&
            customerCreatedMessage.type === 'CustomerCreated' &&
            !!customerCreatedMessage.customer &&
            !!customerCreatedMessage.customer.email &&
            !this.isEventDisabled(this.PROCESSOR_NAME)
        );
    }
    async generateKlaviyoEvents(): Promise<KlaviyoEvent[]> {
        const message = this.ctMessage as unknown as CustomerCreatedMessage;
        logger.info(`processing CT ${message.resource.typeId}${message.type} message`);
        let customer: Customer;
        if ('customer' in message) {
            customer = message.customer;
        } else {
            customer = (await this.context.ctCustomerService.getCustomerProfile(
                (message as CustomerCreatedMessage).resource.id,
            )) as Customer;
        }
        const klaviyoEvents: KlaviyoEvent[] = [];
        const klaviyoEvent: KlaviyoEvent = {
            body: this.context.customerMapper.mapCtCustomerToKlaviyoProfile(customer),
            type: 'profileCreated',
        };
        klaviyoEvents.push(klaviyoEvent);
        logger.info(`Adding klaviyo events for signing up`);
        // todo: very bad way to fetch customer metrics, I need to find a more robust way.
        const body: EventRequest = this.context.customerMapper.mapCtCustomerToKlaviyoEvent(customer, config.get('customer.metrics.customerCreated'));
        const klaviyoCustomerCreatedMetricEvent: KlaviyoEvent = {
            body: body,
            type: 'event'
        }
        klaviyoEvents.push(klaviyoCustomerCreatedMetricEvent);
        return klaviyoEvents;
    }
}
