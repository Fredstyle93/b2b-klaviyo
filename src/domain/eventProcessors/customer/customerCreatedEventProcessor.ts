import { AbstractEvent } from '../abstractEvent';
import logger from '../../../utils/log';
import { CustomerCreatedMessage } from '@commercetools/platform-sdk/dist/declarations/src/generated/models/message';
import { mapCTCustomerToKlaviyoProfile } from './mappers/CTCustomerToKlaviyoProfileMapper';

export class CustomerCreatedEventProcessor extends AbstractEvent {
    isEventValid(): boolean {
        const customerCreatedMessage = this.ctMessage as unknown as CustomerCreatedMessage;
        return (
            customerCreatedMessage.resource.typeId === 'customer' &&
            customerCreatedMessage.type === 'CustomerCreated' &&
            !!customerCreatedMessage.customer &&
            !!customerCreatedMessage.customer.email
        );
    }

    async generateKlaviyoEvents(): Promise<KlaviyoEvent[]> {
        const message = this.ctMessage as unknown as CustomerCreatedMessage;
        logger.info(`processing CT ${message.resource.typeId}${message.type} message`);
        const klaviyoEvent: KlaviyoEvent = {
            body: {
                data: {
                    type: 'profile',
                    attributes: mapCTCustomerToKlaviyoProfile(message.customer),
                },
            },
            type: 'profileCreated',
        };
        return [klaviyoEvent];
    }
}
