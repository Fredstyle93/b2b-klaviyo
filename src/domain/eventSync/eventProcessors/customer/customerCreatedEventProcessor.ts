import { AbstractEventProcessor } from '../abstractEventProcessor';
import logger from '../../../../utils/log';
import { CustomerCreatedMessage } from '@commercetools/platform-sdk/dist/declarations/src/generated/models/message';

export class CustomerCreatedEventProcessor extends AbstractEventProcessor {
    isEventValid(): boolean {
        const customerCreatedMessage = this.ctMessage as unknown as CustomerCreatedMessage;
        return (
            customerCreatedMessage.resource.typeId === 'customer' &&
            customerCreatedMessage.type === 'CustomerCreated' &&
            !!customerCreatedMessage.customer &&
            !!customerCreatedMessage.customer.email
        );
    }

    async generateKlaviyoEvents(): Promise<KlaviyoEvent[]> {
        const message = this.ctMessage as unknown as CustomerCreatedMessage;
        logger.info(`processing CT ${message.resource.typeId}${message.type} message`);
        const klaviyoEvent: KlaviyoEvent = {
            body: this.context.customerMapper.mapCtCustomerToKlaviyoProfile(message.customer),
            type: 'profileCreated',
        };
        return [klaviyoEvent];
    }
}
