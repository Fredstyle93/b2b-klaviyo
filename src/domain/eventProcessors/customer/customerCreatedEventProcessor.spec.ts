import { CustomerCreatedEventProcessor } from './customerCreatedEventProcessor';
import { MessageDeliveryPayload } from '@commercetools/platform-sdk/dist/declarations/src/generated/models/subscription';
import { mockDeep } from 'jest-mock-extended';
import { expect as exp } from 'chai';
import { getSampleCustomerCreatedMessage } from '../../../test/testData/ctCustomerMessages';

describe('customerCreatedEvent > isEventValid', () => {
    it('should return valid when the customer created resource has all the required fields', async () => {
        const ctMessageMock: MessageDeliveryPayload = mockDeep<MessageDeliveryPayload>();
        Object.defineProperty(ctMessageMock, 'resource', { value: { typeId: 'customer' } }); //mock readonly property
        Object.defineProperty(ctMessageMock, 'type', { value: 'CustomerCreated' });
        Object.defineProperty(ctMessageMock, 'customer', { value: { email: 'some@email.com' } });

        const event = CustomerCreatedEventProcessor.instance(ctMessageMock);

        exp(event.isEventValid()).to.be.true;
    });

    it.each`
        resource      | type                 | customer
        ${'invalid'}  | ${'CustomerCreated'} | ${{ email: 'someemail@klaviyo.com' }}
        ${'customer'} | ${'invalid'}         | ${{ email: 'someemail@klaviyo.com' }}
        ${'customer'} | ${'CustomerCreated'} | ${null}
    `('should return invalid when is not a valid customer created message', ({ resource, type, customer }) => {
        const ctMessageMock: MessageDeliveryPayload = mockDeep<MessageDeliveryPayload>();
        Object.defineProperty(ctMessageMock, 'resource', { value: { typeId: resource } });
        Object.defineProperty(ctMessageMock, 'type', { value: type });
        Object.defineProperty(ctMessageMock, 'customer', { value: customer });

        const event = CustomerCreatedEventProcessor.instance(ctMessageMock);

        exp(event.isEventValid()).to.be.false;
    });
});

describe('customerCreatedEvent > generateKlaviyoEvent', () => {
    it('should generate the klaviyo event when the input customer created event is valid', async () => {
        const message = getSampleCustomerCreatedMessage();
        message.customer.addresses.splice(0, message.customer.addresses.length);
        message.customer.addresses.push({
            id: '1235aa3a-5417-4b51-a76c-d6721472531f',
            region: 'aRegion',
            city: 'London',
            country: 'UK',
            phone: '+4407472744666',
            postalCode: 'WE1 2DP',
            streetName: 'High Road',
            streetNumber: '23',
            additionalStreetInfo: 'private access',
            building: 'Tall Tower',
            apartment: 'C',
            additionalAddressInfo: 'additional address info',
            state: 'a state',
        });
        const event = CustomerCreatedEventProcessor.instance(message as unknown as MessageDeliveryPayload);

        const klaviyoEvent = await event.generateKlaviyoEvents();

        exp(klaviyoEvent).to.not.be.undefined;
        exp(klaviyoEvent.length).to.be.eq(1);
        expect(klaviyoEvent[0].body).toMatchSnapshot();
    });

    it('should generate the klaviyo event with the default billing address when set', async () => {
        const message = getSampleCustomerCreatedMessage();
        message.customer.addresses.splice(0, message.customer.addresses.length);
        message.customer.addresses.push({
            id: 'delivery-address-id',
            region: 'aRegion',
            city: 'London',
            country: 'UK',
            phone: '+4407472744666',
            postalCode: 'SE1 5XG',
            streetName: 'some road',
            streetNumber: '23',
            additionalStreetInfo: 'private access',
            building: 'Tall Tower',
            apartment: 'C',
            additionalAddressInfo: 'additional address info',
            state: 'a state',
        });
        message.customer.addresses.push({
            id: 'billing-address-id',
            region: 'aRegion',
            city: 'London',
            country: 'UK',
            phone: '+4407472744666',
            postalCode: 'WS2 5FX',
            streetName: "I'm the default billing address",
            streetNumber: '23',
            additionalStreetInfo: 'private access',
            building: 'Tall Tower',
            apartment: 'C',
            additionalAddressInfo: 'additional address info',
            state: 'a state',
        });
        Object.defineProperty(message, 'customer', {
            value: { ...message.customer, defaultBillingAddressId: 'billing-address-id' },
        }); //mock readonly property
        const event = CustomerCreatedEventProcessor.instance(message as unknown as MessageDeliveryPayload);

        const klaviyoEvent = await event.generateKlaviyoEvents();

        exp(message.customer.addresses.length).to.equal(2);
        exp(klaviyoEvent).to.not.be.undefined;
        exp(klaviyoEvent.length).to.be.eq(1);
        expect(klaviyoEvent[0].body).toMatchSnapshot();
    });

    it('should generate the klaviyo event with the first billing address available when no default billing address is set', async () => {
        const message = getSampleCustomerCreatedMessage();
        message.customer.addresses.splice(0, message.customer.addresses.length);
        message.customer.addresses.push({
            id: 'delivery-address-id',
            region: 'aRegion',
            city: 'London',
            country: 'UK',
            phone: '+4407472744666',
            postalCode: 'SE1 5XG',
            streetName: 'some road',
            streetNumber: '23',
            additionalStreetInfo: 'private access',
            building: 'Tall Tower',
            apartment: 'C',
            additionalAddressInfo: 'additional address info',
            state: 'a state',
        });
        message.customer.addresses.push({
            id: 'first-billing-address-id',
            region: 'aRegion',
            city: 'London',
            country: 'UK',
            phone: '+4407472744666',
            postalCode: 'WS2 5FX',
            streetName: "I'm a billing address",
            streetNumber: '23',
            additionalStreetInfo: 'private access',
            building: 'Tall Tower',
            apartment: 'C',
            additionalAddressInfo: "I'm a billing address",
            state: 'a state',
        });
        message.customer.addresses.push({
            id: 'second-billing-address-id',
            region: 'aRegion',
            city: 'London',
            country: 'UK',
            phone: '+4407472744666',
            postalCode: 'WS2 5FY',
            streetName: "I'm another billing address",
            streetNumber: '23',
            additionalStreetInfo: 'private access',
            building: 'Tall Tower',
            apartment: 'C',
            additionalAddressInfo: "I'm another billing address",
            state: 'a state',
        });

        Object.defineProperty(message, 'customer', {
            value: {
                ...message.customer,
                billingAddressIds: ['second-billing-address-id', 'first-billing-address-id'],
                defaultBillingAddressId: null,
            },
        }); //mock readonly property

        const event = CustomerCreatedEventProcessor.instance(message as unknown as MessageDeliveryPayload);

        const klaviyoEvent = await event.generateKlaviyoEvents();

        exp(message.customer.addresses.length).to.equal(3);
        exp(klaviyoEvent).to.not.be.undefined;
        exp(klaviyoEvent.length).to.be.eq(1);
        expect(klaviyoEvent[0].body).toMatchSnapshot();
    });

    it('should generate the klaviyo event with the first address when no billing address ids are set', async () => {
        const message = getSampleCustomerCreatedMessage();
        message.customer.addresses.splice(0, message.customer.addresses.length);
        message.customer.addresses.push({
            id: 'some-address-id',
            region: 'aRegion',
            city: 'London',
            country: 'UK',
            phone: '+4407472744666',
            postalCode: 'WS2 5FX',
            streetName: "I'm a random address",
            streetNumber: '23',
            additionalStreetInfo: 'private access',
            building: 'Tall Tower',
            apartment: 'C',
            additionalAddressInfo: "I'm a random address",
            state: 'a state',
        });
        message.customer.addresses.push({
            id: 'another-address-id',
            region: 'aRegion',
            city: 'London',
            country: 'UK',
            phone: '+4407472744666',
            postalCode: 'WS2 5FY',
            streetName: "I'm another random address",
            streetNumber: '23',
            additionalStreetInfo: 'private access',
            building: 'Tall Tower',
            apartment: 'C',
            additionalAddressInfo: "I'm another random address",
            state: 'a state',
        });

        Object.defineProperty(message, 'customer', {
            value: {
                ...message.customer,
                billingAddressIds: null,
                defaultBillingAddressId: null,
            },
        }); //mock readonly property

        const event = CustomerCreatedEventProcessor.instance(message as unknown as MessageDeliveryPayload);

        const klaviyoEvent = await event.generateKlaviyoEvents();

        exp(message.customer.addresses.length).to.equal(2);
        exp(klaviyoEvent).to.not.be.undefined;
        exp(klaviyoEvent.length).to.be.eq(1);
        expect(klaviyoEvent[0].body).toMatchSnapshot();
    });

    it('should generate the klaviyo event without customer address inf when no address is available', async () => {
        const message = getSampleCustomerCreatedMessage();
        message.customer.addresses.splice(0, message.customer.addresses.length);
        Object.defineProperty(message, 'customer', {
            value: {
                ...message.customer,
                billingAddressIds: null,
                defaultBillingAddressId: null,
            },
        }); //mock readonly property

        const event = CustomerCreatedEventProcessor.instance(message as unknown as MessageDeliveryPayload);

        const klaviyoEvent = await event.generateKlaviyoEvents();

        exp(message.customer.addresses.length).to.equal(0);
        exp(klaviyoEvent).to.not.be.undefined;
        exp(klaviyoEvent.length).to.be.eq(1);
        expect(klaviyoEvent[0].body).toMatchSnapshot();
    });
});
