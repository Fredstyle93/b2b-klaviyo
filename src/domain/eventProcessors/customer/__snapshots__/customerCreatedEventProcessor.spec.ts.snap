// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`customerCreatedEvent > generateKlaviyoEvent should generate the klaviyo create profile event when the input customer created event is valid 1`] = `
{
  "body": {
    "data": {
      "attributes": {
        "email": "roberto.smith@klaviyo.com",
        "external_id": "e54d8233-be41-4ce0-ae68-5d0674dd8517",
        "first_name": "Roberto",
        "last_name": "Smith",
        "location": {
          "address1": "C, Tall Tower, 23, High Road",
          "address2": "private access, additional address info",
          "city": "London",
          "country": "UK",
          "region": "aRegion",
          "zip": "WE1 2DP",
        },
        "organization": "Klaviyo",
        "phone_number": "+4407476588266",
        "title": "Mr",
      },
      "type": "profile",
    },
  },
  "type": "profileCreated",
}
`;

exports[`customerCreatedEvent > generateKlaviyoEvent should generate the klaviyo create profile event with the default billing address when is available 1`] = `
{
  "body": {
    "data": {
      "attributes": {
        "email": "roberto.smith@klaviyo.com",
        "external_id": "e54d8233-be41-4ce0-ae68-5d0674dd8517",
        "first_name": "Roberto",
        "last_name": "Smith",
        "location": {
          "address1": "C, Tall Tower, 23, I'm the default billing address",
          "address2": "private access, additional address info",
          "city": "London",
          "country": "UK",
          "region": "aRegion",
          "zip": "WS2 5FX",
        },
        "organization": "Klaviyo",
        "phone_number": "+4407476588266",
        "title": "Mr",
      },
      "type": "profile",
    },
  },
  "type": "profileCreated",
}
`;

exports[`customerCreatedEvent > generateKlaviyoEvent should generate the klaviyo create profile event with the first address when no billing address ids are set 1`] = `
{
  "body": {
    "data": {
      "attributes": {
        "email": "roberto.smith@klaviyo.com",
        "external_id": "e54d8233-be41-4ce0-ae68-5d0674dd8517",
        "first_name": "Roberto",
        "last_name": "Smith",
        "location": {
          "address1": "C, Tall Tower, 23, I'm a random address",
          "address2": "private access, I'm a random address",
          "city": "London",
          "country": "UK",
          "region": "aRegion",
          "zip": "WS2 5FX",
        },
        "organization": "Klaviyo",
        "phone_number": "+4407476588266",
        "title": "Mr",
      },
      "type": "profile",
    },
  },
  "type": "profileCreated",
}
`;

exports[`customerCreatedEvent > generateKlaviyoEvent should generate the klaviyo create profile event with the first billing address available when no default billing address is set 1`] = `
{
  "body": {
    "data": {
      "attributes": {
        "email": "roberto.smith@klaviyo.com",
        "external_id": "e54d8233-be41-4ce0-ae68-5d0674dd8517",
        "first_name": "Roberto",
        "last_name": "Smith",
        "location": {
          "address1": "C, Tall Tower, 23, I'm another billing address",
          "address2": "private access, I'm another billing address",
          "city": "London",
          "country": "UK",
          "region": "aRegion",
          "zip": "WS2 5FY",
        },
        "organization": "Klaviyo",
        "phone_number": "+4407476588266",
        "title": "Mr",
      },
      "type": "profile",
    },
  },
  "type": "profileCreated",
}
`;

exports[`customerCreatedEvent > generateKlaviyoEvent should generate the klaviyo create profile event without customer address inf when no address is available 1`] = `
{
  "body": {
    "data": {
      "attributes": {
        "email": "roberto.smith@klaviyo.com",
        "external_id": "e54d8233-be41-4ce0-ae68-5d0674dd8517",
        "first_name": "Roberto",
        "last_name": "Smith",
        "location": null,
        "organization": "Klaviyo",
        "phone_number": null,
        "title": "Mr",
      },
      "type": "profile",
    },
  },
  "type": "profileCreated",
}
`;

